pkg/autoscaler/autoscaler.go

```diff
@@ -365,6 +400,62 @@ func (a *Autoscaler) podCountLimited(desiredPodCount, currentPodCount float64) f
 	return math.Min(desiredPodCount, a.Current().MaxScaleUpRate*currentPodCount)
 }
 
+func (a *Autoscaler) getReadyPods(now time.Time) (float64, error) {
+	if v, ok := a.getReadyPodsFromCache(now); ok {
+		return v, nil
+	}
+
+	if v, err := a.getReadyPodsFromLister(now); err != nil {
+		return 0, err
+	} else {
+		return v, nil
+	}
+}
+
+func (a *Autoscaler) getReadyPodsFromCache(now time.Time) (float64, bool) {
+	a.readyPodsMutex.RLock()
+	defer a.readyPodsMutex.RUnlock()
+
+	v, ok := a.readyPods[now.Round(time.Second)]
+	return v, ok
+}
+
+func (a *Autoscaler) getReadyPodsFromLister(now time.Time) (float64, error) {
+	readyPods := 0
+	endpoints, err := a.endpointsLister.Endpoints(a.namespace).Get(a.revisionService)
+	if errors.IsNotFound(err) {
+		// Treat not found as zero endpoints, it either hasn't been created
+		// or it has been torn down.
+	} else if err != nil {
+		return 0, err
+	} else {
+		for _, es := range endpoints.Subsets {
+			readyPods += len(es.Addresses)
+		}
+	}
+
+	a.readyPodsMutex.Lock()
+	defer a.readyPodsMutex.Unlock()
+
+	// Use 1 as minimum for multiplication and division.
+	float64ReadyPods := math.Max(1, float64(readyPods))
+	a.readyPods[now.Round(time.Second)] = float64ReadyPods
+
+	return float64ReadyPods, nil
+}
+
+func (a *Autoscaler) cleanReadyPodsCache(now time.Time) {
+	a.readyPodsMutex.Lock()
+	defer a.readyPodsMutex.Unlock()
+
+	for key := range a.readyPods {
+		if key.Add(a.Current().StableWindow).Before(now) {
```

Comment:
a.Current() contains a lock inside. 
Do you really need to call it for every item of the readyPods, or you can compute it outside the loop (and the lock as well)?